test/common/quic/envoy_quic_h3_fuzz_test.cc

```diff
@@ -0,0 +1,304 @@
+#include "source/common/quic/envoy_quic_alarm_factory.h"
+#include "source/common/quic/envoy_quic_connection_helper.h"
+#include "source/common/quic/envoy_quic_dispatcher.h"
+#include "source/common/quic/envoy_quic_server_connection.h"
+#include "source/common/quic/envoy_quic_server_session.h"
+
+#include "test/common/quic/envoy_quic_h3_fuzz.pb.h"
+#include "test/common/quic/envoy_quic_h3_fuzz_helper.h"
+#include "test/common/quic/test_proof_source.h"
+#include "test/common/quic/test_utils.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/network/mocks.h"
+
+#include "quiche/quic/core/crypto/null_encrypter.h"
+#include "quiche/quic/core/deterministic_connection_id_generator.h"
+#include "quiche/quic/core/quic_crypto_server_stream.h"
+#include "quiche/quic/core/tls_server_handshaker.h"
+#include "quiche/quic/test_tools/quic_test_utils.h"
+
+namespace Envoy {
+namespace Quic {
+
+using namespace test::common::quic;
+
+// The following classes essentially mock the `QUIC` handshake
+class ProofSourceDetailsSetter {
+public:
+  virtual ~ProofSourceDetailsSetter() = default;
+  virtual void setProofSourceDetails(std::unique_ptr<EnvoyQuicProofSourceDetails> details) = 0;
+};
+
+class TestQuicCryptoServerStream : public quic::QuicCryptoServerStream,
+                                   public ProofSourceDetailsSetter {
+public:
+  ~TestQuicCryptoServerStream() override = default;
+  explicit TestQuicCryptoServerStream(const quic::QuicCryptoServerConfig* crypto_config,
+                                      quic::QuicCompressedCertsCache* compressed_certs_cache,
+                                      quic::QuicSession* session,
+                                      quic::QuicCryptoServerStreamBase::Helper* helper)
+      : quic::QuicCryptoServerStream(crypto_config, compressed_certs_cache, session, helper) {}
+  bool encryption_established() const override { return true; }
+  const EnvoyQuicProofSourceDetails* ProofSourceDetails() const override { return details_.get(); }
+  void setProofSourceDetails(std::unique_ptr<EnvoyQuicProofSourceDetails> details) override {
+    details_ = std::move(details);
+  }
+
```

Comment:
You can override `OnHandshakeMessage()` here which bypasses handshake and calls into session and connection interfaces, i.e.
 SetEncrypter(), SetDefaultEncryptionLevel(), OnNewDecryptionKeyAvailable() and OnDecryptedPacket().

Reply:
The fuzzer does not generate valid handshake messages, hence it will not hit `OnHandshakeMessage()` or `ProcessAdditionalTransportParameters()`, I think.

Reply:
These 2 should be hit as long as initial handshake messages are received, which should be in the corpus, otherwise there won't be any connection created.

Reply:
There are no handshake messages in the corpus. Including a handshake would pose a road block for the fuzzer, because any mutation of the handshake would probably lead to an early error. Since every fuzz case would have to include a valid handshake, this would risk lots of wasted effort.